File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MessageIdGenerator.java
package com.fsck.k9.mail.internet;


import java.util.Locale;
import java.util.UUID;

import android.support.annotation.VisibleForTesting;

import com.fsck.k9.mail.Address;
import com.fsck.k9.mail.Message;


public class MessageIdGenerator {
    public static MessageIdGenerator getInstance() {
        return new MessageIdGenerator();
    }

    @VisibleForTesting
    MessageIdGenerator() {
    }

    public String generateMessageId(Message message) {
        String hostname = null;

        Address[] from = message.getFrom();
        if (from != null && from.length >= 1) {
            hostname = from[0].getHostname();
        }

        if (hostname == null) {
            Address[] replyTo = message.getReplyTo();
            if (replyTo != null && replyTo.length >= 1) {
                hostname = replyTo[0].getHostname();
            }
        }

        if (hostname == null) {
            hostname = "email.android.com";
        }
        
        String uuid = generateUuid();
        return "<" + uuid + "@" + hostname + ">";
    }

    @VisibleForTesting
    protected String generateUuid() {
        // We use upper case here to match Apple Mail Message-ID format (for privacy)
        return UUID.randomUUID().toString().toUpperCase(Locale.US);
    }
}

File: mail/common/src/main/java/com/fsck/k9/mail/internet/MessageIdGenerator.kt
package com.fsck.k9.mail.internet

import com.fsck.k9.mail.Address
import com.fsck.k9.mail.Message
import java.util.UUID

class MessageIdGenerator(private val uuidGenerator: UuidGenerator) {
    fun generateMessageId(message: Message): String {
        val uuid = uuidGenerator.randomUuid()
        val hostname = message.from.firstHostname ?: message.replyTo.firstHostname ?: "fallback.k9mail.app"

        return "<$uuid@$hostname>"
    }

    private val Array<Address>?.firstHostname: String?
        get() = this?.firstOrNull()?.hostname

    companion object {
        @JvmStatic
        fun getInstance(): MessageIdGenerator = MessageIdGenerator(K9UuidGenerator())
    }
}

interface UuidGenerator {
    fun randomUuid(): String
}

class K9UuidGenerator : UuidGenerator {
    override fun randomUuid(): String {
        // We use upper case here to match Apple Mail Message-ID format (for privacy)
        return UUID.randomUUID().toString().uppercase()
    }
}
